### PR TITLE
Ensure TS_VCONN_CLOSE_HOOK hook is called during TS_EVENT_VCONN_CLOSE.

### PR DESCRIPTION
**Pull request by bneradt**
This causes the VCONN_CLOSE hooks to be called more consistently in the
event of TS_EVENT_VCONN_CLOSE.  Without this patch it got missed under
some circumstances, causing situations where plugins would receive a
TS_EVENT_VCONN_START but never a TS_EVENT_VCONN_CLOSE for some
connections.

### CODE REVIEW

**Comment by bneradt** [comment]
[approve ci autest]

File: iocore/net/SSLNetVConnection.cc [hunk 1]
```diff
@@ -1694,6 +1694,12 @@ SSLNetVConnection::callHooks(TSEvent eventId)
   Debug("ssl", "sslHandshakeHookState=%d eventID=%d", this->sslHandshakeHookState, eventId);
 
   // Move state if it is appropriate
+  if (eventId == TS_EVENT_VCONN_CLOSE) {
+    // Regardless of state, if the connection is closing, then transition to
+    // the DONE state. This will trigger us to call the appropriate cleanup
+    // routines.
+    this->sslHandshakeHookState = HANDSHAKE_HOOKS_DONE;
+  }
```
Comment (maskit) [inline iocore/net/SSLNetVConnection.cc:1702]: Looks reasonable. We may want to put the `switch` statement below into `else` clause to ensure `sslHandshakeHookState` won't be accidentally overwritten in the `switch` statement.

Comment (bneradt) [reply iocore/net/SSLNetVConnection.cc:1702]: Good idea.

File: iocore/net/SSLNetVConnection.cc [hunk 2]
```diff
@@ -1694,39 +1694,40 @@ SSLNetVConnection::callHooks(TSEvent eventId)
...
+        this->sslHandshakeHookState = HANDSHAKE_HOOKS_CERT;
+      }
+      break;
+    default:
+      break;
     }
-    break;
-  default:
-    break;
   }
```
Comment (bneradt) [inline iocore/net/SSLNetVConnection.cc:1731]: View this patch using the "Hide whitespace changes" option via github.

**Review by maskit — APPROVED** [review]

**Comment by bneradt** [comment]
Marking this for the 9.1.x train. If plugins rely upon the VCONN_CLOSE hook to free memory, this will result in a memory leak. Or any host of other unexpected behaviors when the hook isn't called when expected.

**Comment by zwoop** [comment]
Cherry-picked to v9.1.x branch.

**Pull request merged**